File: golang/blocking/category_blocking.go
package blocking

import (
	"auction-worker/model"

	"golang.org/x/exp/slices"
)

// 広告とリクエストにあるカテゴリをもとに、配信許可/拒否制御を行う
func CategoryBlocking(advertiser model.Advertiser, category_blocking model.CategoryBlocking, request_category []uint) bool {
	include_ids := false
	for _, category_id := range category_blocking.CategoryIds {
		include_ids = include_ids || slices.Contains(request_category, category_id)
	}
	switch advertiser.CategoryBlockingTypeId {
	case model.ALLOW:
		return !include_ids
	case model.DENY:
		return include_ids
	}
	return false
}

File: python/blocking/category_blocking.py
from model.advertiser import CategoryBlockingType, Advertiser
from model.category_blocking import CategoryBlocking
from typing import List


def categoryBlocking(advertiser: Advertiser, category_blocking: CategoryBlocking, request_category: List[int]) -> bool:
    """広告とリクエストにあるカテゴリをもとに、配信許可/拒否制御を行う

    Args:
        advertiser (Advertiser): 広告主クラス。配信許可/拒否の情報を所持
        category_blocking (CategoryBlocking): 広告主に紐付いているカテゴリ情報
        request_category (List[int]): リクエストに付与されているカテゴリ情報

    Returns:
        bool: カテゴリ制御によってブロックされるときにtrueを返す
    """
    including_ads = False
    for category_id in category_blocking.category_ids:
        including_ads = including_ads or category_id in request_category

    if advertiser.category_blocking_type_id == CategoryBlockingType.ALLOW:
        return not including_ads
    else:
        return including_ads
